lib/MusicBrainz/Server/Data/Role/EntityCache.pm

```diff
@@ -135,43 +150,32 @@ sub _delete_from_cache {
     @ids = uniq grep { defined } @ids;
     return unless @ids;
 
-    my $cache_id = $self->_cache_id;
-
-    # MBS-7241: Lock cache deletions from `_create_cache_entries`
-    # above, so that these keys can't be repopulated until after the
-    # database transaction commits.
-    my @row_ids = uniq
-        # MBS-10497: Limit the number of locks obtained per cache id
-        # per transaction to 50; see the comment in
-        # `_create_cache_entries` above. This increases the chance of
-        # contention, but invalidating cache entries is infrequent
-        # enough that it shouldn't matter.
-        map { $_ % 50 }
-        grep { is_database_row_id($_) } @ids;
-    if (@row_ids) {
-        my $start_time = time;
-        $self->c->sql->do(
-            'SELECT pg_advisory_xact_lock(?, id) ' .
-            '  FROM unnest(?::integer[]) AS id',
-            $cache_id,
-            \@row_ids,
-        );
-        my $elapsed_time = (time - $start_time);
-        if ($elapsed_time > 0.01) {
-            log_debug {
-                "[$start_time] _delete_from_cache: waited $elapsed_time" .
-                ' seconds for ' . (scalar @row_ids) .
-                " locks with cache id $cache_id"
-            };
+    my $c = $self->c;
+    my $cache_prefix = $self->_type . ':';
+
+    if (DBDefs->REPLICATION_TYPE != RT_MIRROR) {
+        my $max_request_time = $c->max_request_time;
+        if (!defined($max_request_time) || $max_request_time == 0) {
+            $max_request_time = $CACHELOCK_TTL_FALLBACK;
+        }
+        my @cachelock_keys = map {
+            [$cache_prefix . 'cachelock:' . $_,
+             1,
+             ($max_request_time + 1)]
+        } @ids;
+        $c->store->set_multi(@cachelock_keys);
+
+        if ($c->sql->is_in_transaction) {
+            $c->sql->add_post_commit_callback(sub {
+                $c->store->delete_multi(@cachelock_keys);
+            });
         }
     }
 
-    my $cache_prefix = $self->_type . ':';
     my @keys = map { $cache_prefix . $_ } @ids;
-
-    my $cache = $self->c->cache($self->_type);
+    my $cache = $c->cache($self->_type);
```

Comment:
What's the point of this being a variable if you don't use it anymore in 178?

Reply:
This might be something I cleaned up locally but didn't push yet. Will make sure it gets fixed if not. 👍